1-Draft/RFCNNNN-OutVariableTransparency.md

```diff
@@ -0,0 +1,40 @@
+---
+RFC:
+Author: Robert Holt
+Status: Draft
+SupercededBy: N/A
+Version:
+Area: Language/Common Variables
+Comments Due: 2018-05-01
+Plan to implement: Yes
+---
+
+# OutVariable Transparency
+
+PowerShell supports OutVariables, which allow users to send the output of a command to the variable rather than to the pipeline.
+
+Since this feature was included in PowerShell v1, the implementation detail of the ArrayList used to collect the output into the OutVariable has been exposed to the user, so that rather than getting the `object` or `object[]` that they would get with pipeline output, an `ArrayList` is always received.
+
+This RFC floats the idea that OutVariable usage should be transparent to the user, so output is agnostic between OutVariable and the pipeline stream. As a corner case in PowerShell's implementation, OutVariables currently represent extra caveats for new users.
+
+## Motivation
+
+    As a PowerShell user, I can use OutVariable just like pipeline output, so that my language experience is simpler and more consistent.
```

Comment:
I disagree with this motivation. 
I specifically use this feature when I do NOT want the default pipeline behavior.

A multitude of cmdlets and functions return objects in an inconsistent way. Because `-OutVariable` always returns an ArrayList, I can ensure the result of the result has a consistent format

Take the following example function as a stand in for get-ADUser and quite a few of the Exchange and Exchange Online cmdlets:

```powershell
function Get-Result {
    [CmdletBinding()]
    param ()
    end {
        :outer
        switch( 1..9 | Get-Random) {
            1 { 'String'; break }
            2 { 1; break }
            3 { $null; break }
            4 { 1,2,3,4; break }
            5 { ,@(1,2,3,4) }
            6 { throw; break }
            7 { Write-Error 'SomeError'; break }
            8 { break :outer }
            9 { break }
        }
    }
}
```

Now, run the following and see all the various issues you encounter:

```powershell
1..20 | %{
    $_
    $Result = 'Stale'
    try {
        $Result = Get-Result
    }
    Catch {
        'Caught'
    }
    finally {
        'Stale? {0}' -f ('Stale' -eq $Result)
        $Result.gettype().name
        $Result.Count
        $Result[0]
    }
}
```

A string will result in the array index being `S`, throw results in the `$result` being stale.

Compare that with 

```powershell
1..20 | %{
    $Result = 'Stale'
    try {
        $null = Get-Result -OutVariable Result
    }
    Catch {
        'Caught'
    }
    finally {
        'Stale? {0}' -f ('Stale' -eq $Result)
        $Result.gettype().name
        $Result.Count
        $Result[0]
    }
}
```

The ArrayList is always created fresh, so it is never stale even when the cmdlet errors or throws. it is always an ArrayList so I don't get weird indexing issues. I can always rely on the methods and properties of an ArrayList to be there.

If this is taken away and change, this will break what I feel is the most stable and reliable method for retrieving output from commands. I will be forced to write ridiculous logic around Microsoft's cmdlets in many cases to try and deal with the insanity of the results then provide.